Feature - Multi Audio

yaroslavgribov opened the pull request:
# Description

This PR introduces new feature - MultiAudio Tracks

## Related PRs

branch | PR
------ | ------
Feature/PLAYER-542 | [Feature/PLAYER-542](https://git.corp.ooyala.com/projects/PBW/repos/video-plugins-internal/pull-requests/297/overview)
Feature/PLAYER-542| [Feature/PLAYER-542](https://github.com/ooyala/skin-config/pull/197)

## Summary

### Components

 - [CloseCaptionMultiAudioMenu Component](https://github.com/ooyala/html5-skin/blob/899bb46d7815b6760731c5b66d04c8c5fc56cea7/js/components/closed-caption-multi-audio-menu/closedCaptionMultiAudioMenu.js) - container component that allows user to select tracks and captions if any available and notifies the controller about selection.

- [Tab Component](https://github.com/ooyala/html5-skin/blob/899bb46d7815b6760731c5b66d04c8c5fc56cea7/js/components/closed-caption-multi-audio-menu/tab.js) - presentational component that displays a list with an icon next to selected item based on design mock-ups.

- [MultiAudioTab Component](https://github.com/ooyala/html5-skin/blob/899bb46d7815b6760731c5b66d04c8c5fc56cea7/js/components/closed-caption-multi-audio-menu/multiAudioTab.js) - presentational component that wraps a `Tab` component and modifies passed `lang` and `label` properties to display user-friendly labels.

## Modules

- [Helpers Module](https://github.com/ooyala/html5-skin/blob/899bb46d7815b6760731c5b66d04c8c5fc56cea7/js/components/closed-caption-multi-audio-menu/helpers.js) - utility functions used in `MultiAudioTab` component

## Adresses 
 -  [As a user, I can see an indicator in the UI that the asset I've chosen has multiaudio tracks.](https://jira.corp.ooyala.com/browse/PLAYER-542)
 - [[WEB] As a user I would like to see userfriendly label for audio track name](https://jira.corp.ooyala.com/browse/PLAYER-3147)

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

- [x] [Tab Component test](https://github.com/ooyala/html5-skin/blob/899bb46d7815b6760731c5b66d04c8c5fc56cea7/tests/components/closed-caption-multi-audio-menu/tab-test.js)
- [x] [MultiAudioTab Component test](https://github.com/ooyala/html5-skin/blob/899bb46d7815b6760731c5b66d04c8c5fc56cea7/tests/components/closed-caption-multi-audio-menu/multiAudioTab-test.js)
- [x] [Helpers Module test](https://github.com/ooyala/html5-skin/blob/899bb46d7815b6760731c5b66d04c8c5fc56cea7/tests/components/closed-caption-multi-audio-menu/helpers-test.js)
- [x] All Unit tests passed without errors

jpfranco commented:
Thanks for the detailed description

jpfranco commented on js/components/closed-caption-multi-audio-menu/closedCaptionMultiAudioMenu.js:
```diff
@@ -0,0 +1,128 @@
+  componentDidMount: function() {
+    var multiAudioCol = ReactDOM.findDOMNode(this.refs.multiAudioCol);
+    var closeCaptionsCol = ReactDOM.findDOMNode(this.refs.closeCaptionsCol);
+    if (multiAudioCol && closeCaptionsCol && typeof Utils.getMaxElementWidth === 'function') {
```
I don't think we need to check `Utils.getMaxElementWidth`, we should be able to trust that our own functions are defined.

yaroslavgribov replied:
Thank you for the review! Good point, I think we can remove that check.

KseniiaSokolovaEpam replied:
\+

pilievOoyala commented on js/components/closed-caption-multi-audio-menu/closedCaptionMultiAudioMenu.js:
```diff
@@ -0,0 +1,128 @@
+  },
+  /**
+   *
+   * @param languages {Array} - list of available languages
```
Can you rename this param to languageList? It will be very easy to make a typo and use the wrong variable and then we'll spend 30 minutes debugging it.

It's also part of the coding standards.

yaroslavgribov replied:
Thank you for the review! Yeah, I think it's better to adjust the naming to follow the standarts, will do.

KseniiaSokolovaEpam replied:
\+

pilievOoyala commented on js/components/closed-caption-multi-audio-menu/closedCaptionMultiAudioMenu.js:
```diff
@@ -0,0 +1,128 @@
+  /**
+   *
+   * @param languages {Array} - list of available languages
+   * @param language {String} - a selected language
```
Change it to: @param language {String} - the selected language

KseniiaSokolovaEpam replied:
\+

pilievOoyala commented on js/components/closed-caption-multi-audio-menu/closedCaptionMultiAudioMenu.js:
```diff
@@ -0,0 +1,128 @@
+
+  /**
+   * when clicking on an item from an cc list, set the corresponding cc value
+   * @param id {string} - id of clicked element
```
Can you add the @public or @private there?

KseniiaSokolovaEpam replied:
\+

pilievOoyala commented on js/components/closed-caption-multi-audio-menu/closedCaptionMultiAudioMenu.js:
```diff
@@ -0,0 +1,128 @@
+
+  /**
+   * when clicking on an item from an audio list, set the corresponding audio value
+   * @param id {string} - id of clicked element
```
Can you add the @public or @private there?

KseniiaSokolovaEpam replied:
\+

jpfranco commented on js/components/closed-caption-multi-audio-menu/closedCaptionMultiAudioMenu.js:
```diff
@@ -0,0 +1,128 @@
+    var closeCaptionsCol = ReactDOM.findDOMNode(this.refs.closeCaptionsCol);
+    if (multiAudioCol && closeCaptionsCol && typeof Utils.getMaxElementWidth === 'function') {
+      var maxWidth = Utils.getMaxElementWidth(multiAudioCol, closeCaptionsCol);
+      multiAudioCol.style.width = maxWidth + 'px';
```
Is this code meant to divide each element into equal size columns? We should be able to do this with flexbox in CSS. Also since this is done in `componentDidMount` it looks like it won't react to style changes.

yaroslavgribov replied:
Good point, we removed that method and replaced it with flexbox implementation.

pilievOoyala commented on js/components/closed-caption-multi-audio-menu/helpers.js:
```diff
@@ -0,0 +1,150 @@
+/**
+ * Gets display label by checking
+ * roles - e.g. nullable field from DASH manifest
+ * and labels - e.g. non-nullable field from HSL manifest
```
Typo, should be HLS and not HSL

KseniiaSokolovaEpam replied:
\+

jpfranco commented on js/components/closed-caption-multi-audio-menu/closedCaptionMultiAudioMenu.js:
```diff
@@ -0,0 +1,128 @@
+      );
+      closeCaptionsCol = (
+        <Tab
+          ref="closeCaptionsCol"
```
This style of `ref` is deprecated, please use a function

KseniiaSokolovaEpam replied:
it was deleted.

aeng7 commented on js/components/closed-caption-multi-audio-menu/closedCaptionMultiAudioMenu.js:
```diff
@@ -0,0 +1,128 @@
+var ClosedCaptionMultiAudioMenu = React.createClass({
+  componentDidMount: function() {
+    var multiAudioCol = ReactDOM.findDOMNode(this.refs.multiAudioCol);
+    var closeCaptionsCol = ReactDOM.findDOMNode(this.refs.closeCaptionsCol);
```
I see closeCaptions instead of closedCaptions in this file. Can we change all these vars/references to closedCaptions? Closed captions is the standard name. This will help maintenance and debugging when searching for closed captions strings, for example.

yaroslavgribov replied:
Good catch, thanks!

KseniiaSokolovaEpam replied:
\+

aeng7 commented on js/components/closed-caption-multi-audio-menu/closedCaptionMultiAudioMenu.js:
```diff
@@ -0,0 +1,128 @@
+   * @private
+   */
+  getClosedCaptions: function(languages, language) {
+    var closeCaptions = [];
```
See above comment on using closedCaptions instead of closeCaptions.

KseniiaSokolovaEpam replied:
\+

jpfranco commented on js/components/closed-caption-multi-audio-menu/helpers.js:
```diff
@@ -0,0 +1,150 @@
+var iso639 = require('iso-639-3');
+var CONSTANTS = require('../../constants/constants');
+var _ = require('underscore');
```
Do we know how much this increases our file size? Underscore was only used in this project as a dev dependency until it was recently introduced with the VR360 updates. I recommend using the ES5 array methods directly (which are supported in all the browsers we support) and importing individual lodash functions as needed. We already bundle underscore in the core.min.js script, so using underscore in this project will add duplicate code

yaroslavgribov replied:
That's a really good comment, thank you! 

Here's `ls -l` output on html5-skin

| filename | With underscore | Without underscore |
| --- | --- | --- |
| html5-skin.js | 2633829 | 2622434 |
| html5-skin.min.js | 1420095 | 1365399 |


It think it's a great idea to use ES5 array methods directly which are supported by the browsers, but the it breaks the tests, because the version of node that we're currently on (I believe it should be somewhere below `0.10.48`) doesn't support these methods in array prototype [link](https://node.green/#ES2015-subclassing-Array-is-subclassable-Array-prototype-filter). I think when we move to more recent version of node we could remove underscore as a dep.

For now I think it could be better to stick with not introducing more dependencies, since underscore is used in [viewControlsVr.js](https://github.com/ooyala/html5-skin/blob/a3ac20b2952bae4e938caf1577c9e3b91f3ea163/js/components/viewControlsVr.js#L5) anyway.

aeng7 commented on js/components/closed-caption-multi-audio-menu/closedCaptionMultiAudioMenu.js:
```diff
@@ -0,0 +1,128 @@
+   * when clicking on an item from an audio list, set the corresponding audio value
+   * @param id {string} - id of clicked element
+   */
+  handleClickMA: function(id) {
```
Can we change this to handleMultiAudioClick? This is more descriptive.

KseniiaSokolovaEpam replied:
\+

aeng7 commented on js/components/closed-caption-multi-audio-menu/closedCaptionMultiAudioMenu.js:
```diff
@@ -0,0 +1,128 @@
+          ref="multiAudioCol"
+          handleClick={this.handleClickMA}
+          skinConfig={this.props.skinConfig}
+          list={this.props.controller.state.multiAudio.tracks}
```
Can we use something more descriptive? List is generic. How about tracks?

yaroslavgribov replied:
Thank you for the review! 
I think `audioTracksList` should work, since it's wrapped in `MultiAudioTab` component.

yaroslavgribov replied:
Changed list to audioTracksList since it's wrapped.

aeng7 commented on js/components/closed-caption-multi-audio-menu/closedCaptionMultiAudioMenu.js:
```diff
@@ -0,0 +1,128 @@
+          ref="closeCaptionsCol"
+          handleClick={this.handleClickCC}
+          skinConfig={this.props.skinConfig}
+          list={closedCaptions}
```
Can we use something more descriptive than "list"? How about captions?

yaroslavgribov replied:
I think `captions` won't work here since it's a generic `Tab` component that wraps both multiAudio and closedCaptions. We could go with `itemsList` thought, what do you think?

yaroslavgribov replied:
Changed to `itemsList`.

aeng7 commented on js/components/closed-caption-multi-audio-menu/closedCaptionMultiAudioMenu.js:
```diff
@@ -0,0 +1,128 @@
+   * when clicking on an item from an cc list, set the corresponding cc value
+   * @param id {string} - id of clicked element
+   */
+  handleClickCC: function(id) {
```
Can we change this to handleClosedCaptionsClick? This is more descriptive.

jpfranco commented on js/components/closed-caption-multi-audio-menu/helpers.js:
```diff
@@ -0,0 +1,150 @@
+  return uniqueTracksList;
+}
+
+module.exports = {
```
Thanks for keeping this code separate, I think it's a great idea to have this code in helper functions :)

jpfranco commented on js/components/closed-caption-multi-audio-menu/tab.js:
```diff
@@ -0,0 +1,69 @@
+        <ul className="oo-cc-ma-menu__list">
+          {this.props.list.map(function(el, index) {
+            return (
+              <li
```
Should this menu items be accessible?

aeng7 commented on js/components/closed-caption-multi-audio-menu/multiAudioTab.js:
```diff
@@ -0,0 +1,54 @@
+
+var MultiAudioTab = React.createClass({
+  render: function() {
+    var audioTracksList = this.props.list.map(
```
Is this something we can do in componentWillUpdate/componentDidUpdate? So we don't recreate this list every time the component renders.

yaroslavgribov replied:
Yeah it could be a good idea, 
but component gets unmounted after we close the menu, and lifecycle methods won't be called, because it's a new render every time we open the menu. We could work around it with static fields, but it'd create unnecessary complexity in checking for updates, so I think it's not worth it overall.

aeng7 replied:
Sounds good

jpfranco commented on js/components/utils.js:
```diff
@@ -953,6 +953,27 @@ var Utils = {
   },
+
+  /**
+   * the function returns the maximum width of the elements
```
I posted earlier that we should look into doing this with CSS, but in case this needed, we should add a more detailed description. It's not clear what this maximum width is or what it is used for.

KseniiaSokolovaEpam replied:
\+

aeng7 commented on js/components/utils.js:
```diff
@@ -953,6 +953,27 @@ var Utils = {
+   */
+  getMaxElementWidth: function() {
+    var argumentsArray = Array.prototype.slice.call(arguments);
+    var startWidth = 0;
```
Should we use maxWidth instead of startWidth, since the function is returning the max element width?

KseniiaSokolovaEpam replied:
it was removed.

aeng7 commented on js/constants/constants.js:
```diff
@@ -106,7 +106,10 @@ module.exports = {
+    SELECT_TO_UNMUTE: "SELECT TO UNMUTE",
+    AUDIO: "Audio",
+    SUBTITLES: "Subtitles",
+    UNDEFINED_LANGUAGE: "Undefined language"
```
Have we checked with Steve Paddon and Fernando I. about this string?

yaroslavgribov replied:
Yeah, there's a [question](https://wiki.corp.ooyala.com/pages/viewpage.action?spaceKey=ENG&title=Supporting+Audio+Service+Type+in+OTS#link-talk-3468) in wiki on the topic.

aeng7 commented on js/controller.js:
```diff
@@ -298,6 +298,8 @@ OO.plugin("Html5Skin", function (OO, _, $, W) {
 
       this.state.showMultiAudioIcon = !!params.showMultiAudioIcon;
+      //ToDo: it's needs to test
+      // this.state.showMultiAudioIcon = true;
```
Do we still need this commented line?

KseniiaSokolovaEpam replied:
\+

aeng7 commented on js/skin.js:
```diff
@@ -432,6 +434,22 @@ var Skin = React.createClass({
+          screen = (
+            <ContentScreen
+              {...this.props}
+              cssClass="oo-close-button oo-close-button--ma"
```
Should there be two dashes ```--``` before the ```ma```?

yaroslavgribov replied:
Yeah I think it's a part BEM (modifier).

pilievOoyala commented on js/components/controlBar.js:
```diff
@@ -236,8 +236,12 @@ var ControlBar = React.createClass({
 
-    this.togglePopover(CONSTANTS.MENU_OPTIONS.MULTI_AUDIO);
-    this.closePopovers();
+    if (this.props.responsiveView === this.props.skinConfig.responsive.breakpoints.lg.id) {
```
I think we need some null checks for this variable
this.props.skinConfig.responsive.breakpoints.lg.id
Props and skinConfig are guaranteed to exist but I don't know about the rest.

pilievOoyala replied:
So make it something like this:

if (this.props.skinConfig.resposive &&
this.props.skinConfig.responsive.breakpoints &&
this.props.skinConfig.responsive.breakpoints.lg && 
(this.props.responsiveView === this.props.skinConfig.responsive.breakpoints.lg.id) ){
//do stuff
}

pilievOoyala replied:
@jpfranco @aeng7  Correct me if I'm wrong.

KseniiaSokolovaEpam replied:
\+

jpfranco commented on tests/components/controlBar-test.js:
```diff
@@ -951,44 +951,7 @@ describe('ControlBar', function () {
     expect(ccButtons.length).toBe(1);
   });
-
-  it('shows/hides multiaudio button if multiaudio available', function() {
```
Do we need a replacement for this test or this button now always shown?

pilievOoyala commented on js/components/icon.js:
```diff
@@ -4,7 +4,7 @@ var React = require('react'),
 var Icon = React.createClass({
   shouldComponentUpdate: function(nextProps, nextState) {
-    return  (nextProps.icon != this.props.icon);
+    return  (nextProps.icon != this.props.icon || nextProps.className !== this.props.className);
```
can the first part be changed to '!==' ?

pilievOoyala replied:
If not can you leave a comment saying why.  We are trying to move all the old code always use === and !== if possible.

KseniiaSokolovaEpam replied:
\+

KseniiaSokolovaEpam replied:
\+

jpfranco commented on tests/components/multi-audio/multiAudioPopover-test.js:
```diff
@@ -1,53 +0,0 @@
-jest.dontMock('../../../js/components/multi-audio/multiAudioPopover');
```
Are we missing tests for `multiAudioCloseCaptionScreen` or other new components?

KseniiaSokolovaEpam replied:
\+

yaroslavgribov replied:
[Added tests for closedCaptionsMultiAudioMenu component](https://github.com/ooyala/html5-skin/pull/1035/commits/7aad5e2443f372b6745bf3006179956400646d5c)

pilievOoyala commented on js/components/utils.js:
```diff
@@ -953,6 +953,27 @@ var Utils = {
+  /**
+   * the function returns the maximum width of the elements
+   * @param {...Element} DOM Element
+   * @returns {number} - max width of the elements
```
Can you add in the @private or @public tag? I don't know which it's supposed to be.

KseniiaSokolovaEpam replied:
It was removed

pilievOoyala commented on js/constants/constants.js:
```diff
@@ -106,7 +106,10 @@ module.exports = {
     SHARE: "Share",
     MORE_OPTIONS: "More Options",
-    SELECT_TO_UNMUTE: "SELECT TO UNMUTE"
+    SELECT_TO_UNMUTE: "SELECT TO UNMUTE",
```
Can you change this to 'Select to unmute', so it's consistent with the other variables?

aeng7 replied:
This is the string we see for the unmute icon. The intent is for them to be all upper case.

pilievOoyala replied:
Ok, sounds good to me.

aeng7 commented:
@yaroslavgribov , have we tested the new fonts in IE/Edge? We've had some issues in the past on those browsers with font updates.

pilievOoyala commented on js/constants/constants.js:
```diff
@@ -254,7 +257,8 @@ module.exports = {
     CHINESE: "zh",
-    JAPANESE: "ja"
+    JAPANESE: "ja",
+    NOT_MATCHED: "und"
```
@jpfranco @aeng7 Do we have a different value that we use when it's undefined usually?  I remember something about 'zz' but maybe that's not related to this.

jpfranco replied:
Hmm, not sure. I think we also need to add Korean though

yaroslavgribov replied:
There's a section in [ISO-639-2 faq](https://www.loc.gov/standards/iso639-2/faq.html#25) and it actually mentions  `zxx (No linguistic content; Not applicable)`, but I think our case is more about
> If there is language content, but the specific language cannot be determined a special identifier is provided by ISO 639-2  **und (Undetermined)**

Although I think it could be better to change the constant to `UNDETERMINED`.

jpfranco commented:
Another thing to consider is that we recently implemented accessibility support for the closed captions menu for an important customer. If this new menu will replace our current CC menu then we will most likely need to make it fully accessible. Please make sure to discuss this with Steve Paddon

KseniiaSokolovaEpam commented on js/controller.js:
```diff
@@ -1749,6 +1752,24 @@ OO.plugin("Html5Skin", function (OO, _, $, W) {
+    toggleMultiAudio: function () {
+      if (this.state.screenToShow == CONSTANTS.SCREEN.MULTI_AUDIO_SCREEN) {
+        this.closeScreen();
+        this.renderSkin();
```
Ok, I have a task about toggleMultiAudio functionality. So It would be very comfortable for me to change it in next PR. Thank you! I will remember about your comment.

yaroslavgribov commented:
Thank you everyone for thorough review! 
We've addressed most of the comments made in this PR and added more tests for the functionality. Please have a look at the updated PR.

jpfranco commented on assets/fonts/ooyala-slick-type.svg:
```diff
@@ -2,7 +2,7 @@
 <svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1">
 <metadata>
-Created by FontForge 20170730 at Sun Dec  3 16:46:33 2017
+Created by FontForge 20170730 at Fri Mar 16 16:00:39 2018
```
These icons still have the issue brought up in https://github.com/ooyala/html5-skin/pull/1024. Please check to see if regenerating the icons helps with this problem.

jpfranco commented on js/components/closed-caption-multi-audio-menu/closedCaptionMultiAudioMenu.js:
```diff
@@ -0,0 +1,115 @@
+var Tab = require('./tab');
+var MultiAudioTab = require('./multiAudioTab');
+
+var ClosedCaptionMultiAudioMenu = React.createClass({
```
How do we handle the old Closed Captions menu now that we have this screen? Is the old menu disabled when multi audio is available?

yaroslavgribov replied:
I think both menus are available at this moment and they both work interchangeably, it's just that the old one provides far more customisation options (color, opacity, font size etc.), while the new one provides only language selection mechanism.

jpfranco commented:
@yaroslavgribov we discussed the accessibility requirement during standup today and we agreed that we would like to make all new components accessible moving forward. I will file a new ticket for this.

pilievOoyala commented:
LGTM. But need one more approval. @jpfranco can you merge if you approve.

jpfranco commented:
I'm ok to merge but please look into the icon issue for a future PR @yaroslavgribov @KseniiaSokolovaEpam. I think @aeng7 was going to take one last look

aeng7 commented:
yep LGTM
